sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/SelectQueryFunctionTest.kt

```diff
@@ -429,6 +434,258 @@ class SelectQueryFunctionTest {
       """.trimMargin())
   }
 
+  @Test fun `types are exposed properly in HSQL`(dialect: DialectPreset) {
+    assumeTrue(dialect == DialectPreset.HSQL)
+
+    val file = FixtureCompiler.parseSql("""
+      |CREATE TABLE data (
+      |  boolean0 BOOLEAN NOT NULL,
+      |  boolean1 BOOLEAN,
+      |  boolean2 BOOLEAN AS kotlin.String NOT NULL,
+      |  boolean3 BOOLEAN AS kotlin.String,
+      |  tinyint0 TINYINT NOT NULL,
+      |  tinyint1 TINYINT,
+      |  tinyint2 TINYINT AS kotlin.String NOT NULL,
+      |  tinyint3 TINYINT AS kotlin.String,
+      |  smallint0 SMALLINT NOT NULL,
+      |  smallint1 SMALLINT,
+      |  smallint2 SMALLINT AS kotlin.String NOT NULL,
+      |  smallint3 SMALLINT AS kotlin.String,
+      |  int0 ${dialect.intType} NOT NULL,
+      |  int1 ${dialect.intType},
+      |  int2 ${dialect.intType} AS kotlin.String NOT NULL,
+      |  int3 ${dialect.intType} AS kotlin.String,
+      |  bigint0 BIGINT NOT NULL,
+      |  bigint1 BIGINT,
+      |  bigint2 BIGINT AS kotlin.String NOT NULL,
+      |  bigint3 BIGINT AS kotlin.String
+      |);
+      |
+      |selectData:
+      |SELECT *
+      |FROM data;
+      """.trimMargin(), tempFolder, dialectPreset = dialect)
+
+    val query = file.namedQueries.first()
+    val generator = SelectQueryGenerator(query)
+
+    assertThat(generator.customResultTypeFunction().toString()).isEqualTo("""
+      |public override fun <T : kotlin.Any> selectData(mapper: (
+      |  boolean0: kotlin.Boolean,
+      |  boolean1: kotlin.Boolean?,
+      |  boolean2: kotlin.String,
+      |  boolean3: kotlin.String?,
+      |  tinyint0: kotlin.Byte,
+      |  tinyint1: kotlin.Byte?,
+      |  tinyint2: kotlin.String,
+      |  tinyint3: kotlin.String?,
+      |  smallint0: kotlin.Short,
+      |  smallint1: kotlin.Short?,
+      |  smallint2: kotlin.String,
+      |  smallint3: kotlin.String?,
+      |  int0: kotlin.Int,
+      |  int1: kotlin.Int?,
+      |  int2: kotlin.String,
+      |  int3: kotlin.String?,
+      |  bigint0: kotlin.Long,
+      |  bigint1: kotlin.Long?,
+      |  bigint2: kotlin.String,
+      |  bigint3: kotlin.String?
+      |) -> T): com.squareup.sqldelight.Query<T> = com.squareup.sqldelight.Query(${query.id}, selectData, driver, "Test.sq", "selectData", ""${'"'}
+      ||SELECT *
+      ||FROM data
+      |""${'"'}.trimMargin()) { cursor ->
+      |  mapper(
+      |    cursor.getLong(0)!! == 1L,
+      |    cursor.getLong(1)?.let { it == 1L },
+      |    database.data_Adapter.boolean2Adapter.decode(cursor.getLong(2)!! == 1L),
+      |    cursor.getLong(3)?.let { database.data_Adapter.boolean3Adapter.decode(it == 1L) },
+      |    cursor.getLong(4)!!.toByte(),
+      |    cursor.getLong(5)?.toByte(),
```

Comment:
With `DialectType#decode(…)` this should actually be `cursor.getLong(5)?.let { it.toByte() }`. The reason it is like this is that the default adapters are intercepting it before `DialectType#decode(…)` can. I think this is fine atm, as we're planning on removing the default adapters (#2056) and the generated code is basically the same.